pKiss segfaults with arbitrary sequence.

Running `pKiss --mode mfe --strategy P` crashes with a segmentation fault after entering an arbitrary RNA sequence.
Equivalently `pKiss_mfe -s P <RNA sequence>` produces the same crash. 
As far as I can tell, the strategy doesn't matter.
Tough the crash happens in `gapc` code, it appears to be related to extension code of `fold-grammars` (cf. `Extensions/pknot_shape.hh`).
I remember having no problems back in May or so. Might be related to my `glibc` version but I have no clue about this.
I compiled `gapc` ([current master](https://github.com/jlab/gapc/commit/10f06d31517436dcd2ffbe39b0add221eddbe15b)) with `GCC 10.2.0` and `glibc 2.32`:
```
CXXFLAGS="-fpermissive -ggdb -fno-eliminate-unused-debug-symbols -O0" ./configure --enable---debug && make
```

`-fpermissive` was necessary to make `gcc` ignore:
```
src/lexer.l: in function »yy::Parser::token_type yylex(yy::Parser::semantic_type*, yy::Parser::location_type*, yy::Parser::token_type)«:
src/lexer.l:210:52: error: invalid cast from »yy::position::filename_type*« {aka »const std::__cxx11::basic_string<char>*«} to »std::string*« {aka »std::__cxx11::basic_string<char>*«} [-fpermissive]
  210 | 
      |                                                    ^                                          
      |                                                    |
      |                                                    yy::position::filename_type* {aka const std::__cxx11::basic_string<char>*}
make: *** [makefiles/deps.mf:33: src/lexer.o] error 1
```

I'll open a separate issue in the `gapc` repository if you prefer it that way.
`pKiss` ([current master](https://github.com/jlab/fold-grammars/commit/0b3c311ea1d6455d3b2b71f35703c5d76e9d029a)) was compiled with
```
CXXFLAGS_EXTRA="-O0 -ggdb3 -fno-eliminate-unused-debug-symbols -DDEBUG"
```
to get the necessary debugging symbols.

Here is a backtrace from `gdb` (paths shortened):
```
Starting program: fold-grammars/Misc/Applications/pKiss/x86_64-pc-linux-gnu/pKiss_mfe -s P ACGUCCGA

Program received signal SIGSEGV, Segmentation fault.
0x000055555556f459 in Map::Pool<unsigned long, 6553600ul>::free (this=0x5550003699fc, 
    t=0x7ffff1298018)
    at /usr/include/rtlib/map_pool.hh:306
306	      ret->set_next(head);
(gdb) bt
#0  0x000055555556f459 in Map::Pool<unsigned long, 6553600ul>::free (this=0x5550003699fc, 
    t=0x7ffff1298018)
    at /usr/include/rtlib/map_pool.hh:306
#1  0x000055555556e529 in MultiPool<unsigned long>::free (
    this=0x55555562a700 <Fiber<unsigned long, unsigned char, PkAlph<unsigned long, unsigned char> >::pool>, x=0x7ffff1298018, n=1)
    at /usr/include/rtlib/multipool.hh:98
#2  0x000055555556cdd8 in Fiber<unsigned long, unsigned char, PkAlph<unsigned long, unsigned char> >::dealloc (this=0x55555562a6e8 <closeParen>, t=0x7ffff1298018, n=1 '\001')
    at /usr/include/rtlib/shape.hh:68
#3  0x000055555556c065 in Fiber<unsigned long, unsigned char, PkAlph<unsigned long, unsigned char> >::~Fiber (this=0x55555562a6e8 <closeParen>, __in_chrg=<optimized out>)
    at /usr/include/rtlib/shape.hh:121
#4  0x00007ffff76dadb7 in __run_exit_handlers () from /usr/lib64/libc.so.6
#5  0x00007ffff76daf5e in exit () from /usr/lib64/libc.so.6
#6  0x00007ffff76c3159 in __libc_start_main () from /usr/lib64/libc.so.6
#7  0x000055555556463e in _start ()
```
Let me know if you need any additional information, I'd be happy to provide it.

--- a/.travis.yml
+++ b/.travis.yml
@@ -11,7 +11,7 @@ env:
 before_install:
   - sudo apt-get update -y
   - sudo apt-get install flex bison make libboost-all-dev libgsl-dev ghc
-  - git clone --branch codestyle https://github.com/jlab/gapc /home/travis/gapc
+  - git clone --branch master https://github.com/jlab/gapc /home/travis/gapc
   - pushd .
   - cd /home/travis/gapc
   - ./configure

--- a/.travis.yml
+++ b/.travis.yml
@@ -11,7 +11,7 @@ env:
 before_install:
   - sudo apt-get update -y
   - sudo apt-get install flex bison make libboost-all-dev libgsl-dev ghc
-  - git clone https://github.com/jlab/gapc /home/travis/gapc
+  - git clone --branch codestyle https://github.com/jlab/gapc /home/travis/gapc
   - pushd .
   - cd /home/travis/gapc
   - ./configure

--- a/Algebras/Shapes/Parts/algpart_shape1_basic.gap
+++ b/Algebras/Shapes/Parts/algpart_shape1_basic.gap
@@ -1,4 +1,4 @@
-//openParen and closeParen are defined in Extensions/shapes.hh as '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
+//openParen and closeParen are defined in Extensions/shapes.hh as char '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
 
   shape_t sadd(Subsequence b, shape_t e) {
     if (front(e) == '_') {
@@ -28,30 +28,30 @@
   }
 
   shape_t bl(Subsequence lb,Subsequence lregion,shape_t e,Subsequence rb) {
-    return openParen + '_' + e + closeParen;
+    return shape_t(openParen) + '_' + e + shape_t(closeParen);
   }
 
   shape_t br(Subsequence lb,shape_t e,Subsequence rregion,Subsequence rb) {
-    return openParen + e + '_' + closeParen;
+    return shape_t(openParen) + e + '_' + shape_t(closeParen);
   }
 
   shape_t il(Subsequence lb,Subsequence lregion,shape_t e,Subsequence rregion,Subsequence rb) {
-    return openParen + '_' + e + '_' + closeParen;
+    return shape_t(openParen) + '_' + e + '_' + shape_t(closeParen);
   }
 
   shape_t mldr(Subsequence lb,shape_t e,Subsequence dr,Subsequence rb) {
     if (back(e) == '_') {
-      return openParen + e + closeParen;
+      return shape_t(openParen) + e + shape_t(closeParen);
     } else {
-      return openParen + e + shape_t('_') + closeParen; //cannot happen in macrostates, because this is handled in the mladr case
+      return shape_t(openParen) + e + shape_t('_') + shape_t(closeParen); //cannot happen in macrostates, because this is handled in the mladr case
     }
   }
 
   shape_t mldl(Subsequence lb,Subsequence dl,shape_t e,Subsequence rb) {
     if (front(e) == '_') {
-      return openParen + e + closeParen;
+      return shape_t(openParen) + e + shape_t(closeParen);
     } else {
-      return openParen + shape_t('_') + e + closeParen; //cannot happen in macrostates, because this is handled in the mladl case
+      return shape_t(openParen) + shape_t('_') + e + shape_t(closeParen); //cannot happen in macrostates, because this is handled in the mladl case
     }
   }
 
@@ -65,9 +65,9 @@
     if (back(res) != '_') {
       res = res + shape_t('_'); //cannot happen in macrostates
     }
-    return openParen + res + closeParen;
+    return shape_t(openParen) + res + shape_t(closeParen);
   }
-  
+
   shape_t addss(shape_t x,Subsequence rb) {
     if (back(x) == '_') {
       return x;

--- a/Algebras/Shapes/Parts/algpart_shape2_basic.gap
+++ b/Algebras/Shapes/Parts/algpart_shape2_basic.gap
@@ -1,13 +1,13 @@
-//openParen and closeParen are defined in Extensions/shapes.hh as '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
+//openParen and closeParen are defined in Extensions/shapes.hh as char '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
 
   shape_t bl(Subsequence lb,Subsequence lregion,shape_t e,Subsequence rb) {
-    return openParen + '_' + e + closeParen;
+    return shape_t(openParen) + '_' + e + shape_t(closeParen);
   }
 
   shape_t br(Subsequence lb,shape_t e,Subsequence rregion,Subsequence rb) {
-    return openParen + e + '_' + closeParen;
+    return shape_t(openParen) + e + '_' + shape_t(closeParen);
   }
 
   shape_t il(Subsequence lb,Subsequence lregion,shape_t e,Subsequence rregion,Subsequence rb) {
-    return openParen + '_' + e + '_' + closeParen;
+    return shape_t(openParen) + '_' + e + '_' + shape_t(closeParen);
   }

--- a/Algebras/Shapes/Parts/algpart_shape3_basic.gap
+++ b/Algebras/Shapes/Parts/algpart_shape3_basic.gap
@@ -1,13 +1,13 @@
-//openParen and closeParen are defined in Extensions/shapes.hh as '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
+//openParen and closeParen are defined in Extensions/shapes.hh as char '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
 
   shape_t bl(Subsequence lb,Subsequence lregion,shape_t e,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }
 
   shape_t br(Subsequence lb,shape_t e,Subsequence rregion,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }
-  
+
   shape_t il(Subsequence lb,Subsequence lregion,shape_t e,Subsequence rregion,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }

--- a/Algebras/Shapes/Parts/algpart_shape4_basic.gap
+++ b/Algebras/Shapes/Parts/algpart_shape4_basic.gap
@@ -1,5 +1,5 @@
-//openParen and closeParen are defined in Extensions/shapes.hh as '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
+//openParen and closeParen are defined in Extensions/shapes.hh as char '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
 
   shape_t il(Subsequence lb,Subsequence lregion,shape_t e,Subsequence rregion,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }

--- a/Algebras/Shapes/Parts/algpart_shape5_basic.gap
+++ b/Algebras/Shapes/Parts/algpart_shape5_basic.gap
@@ -1,8 +1,8 @@
-//openParen and closeParen are defined in Extensions/shapes.hh as '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
+//openParen and closeParen are defined in Extensions/shapes.hh as char '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
 
   shape_t sadd(Subsequence b, shape_t e) {
     shape_t emptyShape;
-    
+
     if (e == emptyShape) {
       return '_' + e;
     } else {
@@ -48,7 +48,7 @@
   }
 
   shape_t hl(Subsequence lb,Subsequence region,Subsequence rb) {
-    return openParen + closeParen;
+    return shape_t(openParen) + shape_t(closeParen);
   }
 
 
@@ -65,23 +65,23 @@
   }
 
   shape_t ml(Subsequence lb,shape_t e,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }
 
   shape_t mlall(Subsequence lb,shape_t e,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }
 
   shape_t mldr(Subsequence lb,shape_t e,Subsequence dr,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }
 
   shape_t mldlr(Subsequence lb,Subsequence dl,shape_t e,Subsequence dr,Subsequence rb) {
-    return openParen + e+ closeParen;
+    return shape_t(openParen) + e+ shape_t(closeParen);
   }
 
   shape_t mldl(Subsequence lb,Subsequence dl,shape_t e,Subsequence rb) {
-    return openParen + e+ closeParen;
+    return shape_t(openParen) + e+ shape_t(closeParen);
   }
 
   shape_t addss(shape_t e,Subsequence rb) {

--- a/Algebras/Shapes/Parts/algpart_shapeX_basic.gap
+++ b/Algebras/Shapes/Parts/algpart_shapeX_basic.gap
@@ -1,8 +1,8 @@
-//openParen and closeParen are defined in Extensions/shapes.hh as '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
+//openParen and closeParen are defined in Extensions/shapes.hh as char '[' ']' and in Extensions/pknot_shape.hh as '(' ')'
 
   shape_t sadd(Subsequence b, shape_t x) {
     shape_t emptyShape;
-    
+
     if (x == emptyShape) {
       return shape_t('_');
     } else {
@@ -26,7 +26,7 @@
 		  return le;
 		} else {
 		  return le + re;
-		}			
+		}
 	}
   }
 
@@ -72,17 +72,17 @@
   }
 
   shape_t hl(Subsequence lb,Subsequence region,Subsequence rb) {
-    return openParen + closeParen;
+    return shape_t(openParen) + shape_t(closeParen);
   }
 
 
   shape_t bl(Subsequence lb,Subsequence lregion,shape_t x,Subsequence rb) {
 	if (shapelevel() <= 3) {
 		shape_t res;
-		append(res, openParen);
+		append(res, shape_t(openParen));
 		if (shapelevel() <= 2) { append(res, '_'); }
 		append(res, x);
-		append(res, closeParen);
+		append(res, shape_t(closeParen));
 		return res;
 	} else {
 		return x;
@@ -92,10 +92,10 @@
   shape_t br(Subsequence lb,shape_t x,Subsequence rregion,Subsequence rb) {
 	if (shapelevel() <= 3) {
 		shape_t res;
-		append(res, openParen);
+		append(res, shape_t(openParen));
 		append(res, x);
 		if (shapelevel() <= 2) { append(res, '_'); }
-		append(res, closeParen);
+		append(res, shape_t(closeParen));
 		return res;
 	} else {
 		return x;
@@ -105,11 +105,11 @@
   shape_t il(Subsequence lb,Subsequence lregion,shape_t x,Subsequence rregion,Subsequence rb) {
 	if (shapelevel() <= 4) {
 		shape_t res;
-		append(res, openParen);
+		append(res, shape_t(openParen));
 		if (shapelevel() <= 2) { append(res, '_'); }
 		append(res, x);
 		if (shapelevel() <= 2) { append(res, '_'); }
-		append(res, closeParen);
+		append(res, shape_t(closeParen));
 		return res;
 	} else {
 		return x;
@@ -117,39 +117,39 @@
   }
 
   shape_t ml(Subsequence lb,shape_t e,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }
 
   shape_t mlall(Subsequence lb,shape_t e,Subsequence rb) {
-    return openParen + e + closeParen;
+    return shape_t(openParen) + e + shape_t(closeParen);
   }
 
   shape_t mldr(Subsequence lb,shape_t e,Subsequence dr,Subsequence rb) {
 	  shape_t res;
-	  append(res, openParen);
+	  append(res, shape_t(openParen));
 	  append(res, e);
 	  if ((shapelevel() == 1) && (back(e) != '_')) { append(res, '_'); }
-	  append(res, closeParen);
+	  append(res, shape_t(closeParen));
 	  return res;
   }
 
   shape_t mldlr(Subsequence lb,Subsequence dl,shape_t e,Subsequence dr,Subsequence rb) {
 	  shape_t res;
-	  append(res, openParen);
+	  append(res, shape_t(openParen));
 	  if ((shapelevel() == 1) && (front(e) != '_')) { append(res, '_'); }
 	  append(res, e);
 	  if ((shapelevel() == 1) && (back(e) != '_')) { append(res, '_'); }
-	  append(res, closeParen);
+	  append(res, shape_t(closeParen));
 	  return res;
-	  
+
   }
 
   shape_t mldl(Subsequence lb,Subsequence dl,shape_t e,Subsequence rb) {
 	  shape_t res;
-	  append(res, openParen);
+	  append(res, shape_t(openParen));
 	  if ((shapelevel() == 1) && (front(e) != '_')) { append(res, '_'); }
 	  append(res, e);
-	  append(res, closeParen);
+	  append(res, shape_t(closeParen));
 	  return res;
   }
 

--- a/Extensions/pknot_shape.hh
+++ b/Extensions/pknot_shape.hh
@@ -1,5 +1,5 @@
-#ifndef RTLIB_SHAPE_HH_  // SMJ 2020-11-14: don't know why, but must be identical to FLAG in gapc/rtlib/shape.hh
-#define RTLIB_SHAPE_HH_
+#ifndef PKNOT_SHAPE_HH
+#define PKNOT_SHAPE_HH
 
 template<typename alphabet, typename pos_type, typename T>
 inline bool midsize(const Basic_Sequence<alphabet, pos_type> &seq, T i, T j, int a, int l) {
@@ -110,7 +110,7 @@ inline SHAPE tail(const SHAPE &a) {
 
 typedef shape_t myShape;
 
-static const shape_t openParen = '(';
-static const shape_t closeParen = ')';
+static const char openParen = '(';
+static const char closeParen = ')';
 
 #endif

--- a/Extensions/pknot_shape.hh
+++ b/Extensions/pknot_shape.hh
@@ -1,5 +1,5 @@
-#ifndef PKNOT_SHAPE_HH
-#define PKNOT_SHAPE_HH
+#ifndef RTLIB_SHAPE_HH_  // SMJ 2020-11-14: don't know why, but must be identical to FLAG in gapc/rtlib/shape.hh
+#define RTLIB_SHAPE_HH_
 
 template<typename alphabet, typename pos_type, typename T>
 inline bool midsize(const Basic_Sequence<alphabet, pos_type> &seq, T i, T j, int a, int l) {

--- a/Extensions/shapes.hh
+++ b/Extensions/shapes.hh
@@ -1,5 +1,5 @@
-#ifndef SHAPE_HH_
-#define SHAPE_HH_
+#ifndef SHAPES_HH
+#define SHAPES_HH
 
 static const char openParen = '[';
 static const char closeParen = ']';

--- a/Extensions/shapes.hh
+++ b/Extensions/shapes.hh
@@ -1,7 +1,7 @@
-#ifndef RTLIB_SHAPE_HH_  // SMJ 2020-11-14: don't know why, but must be identical to FLAG in gapc/rtlib/shape.hh
-#define RTLIB_SHAPE_HH_
+#ifndef SHAPE_HH_
+#define SHAPE_HH_
 
-static const shape_t openParen = '[';
-static const shape_t closeParen = ']';
+static const char openParen = '[';
+static const char closeParen = ']';
 
 #endif

--- a/Extensions/shapes.hh
+++ b/Extensions/shapes.hh
@@ -1,4 +1,4 @@
-#ifndef RTLIB_SHAPE_HH_  # SMJ: don't know why, but must be identical to FLAG in gapc/rtlib/shape.hh
+#ifndef RTLIB_SHAPE_HH_  // SMJ 2020-11-14: don't know why, but must be identical to FLAG in gapc/rtlib/shape.hh
 #define RTLIB_SHAPE_HH_
 
 static const shape_t openParen = '[';

--- a/Misc/Applications/lib/foldGrammars/Settings.pm
+++ b/Misc/Applications/lib/foldGrammars/Settings.pm
@@ -6,18 +6,18 @@ use warnings;
 package Settings;
 
 our %PROGINFOS = (
-	'rnashapes', 				{date => '01.10.2015', version => '3.3.0', name => 'RNAshapes', packageDir => 'RNAshapes/'},
-	'rnaalishapes', 			{date => '20.03.2015', version => '2.4.7', name => 'RNAalishapes', packageDir => 'RNAalishapes/'},
-	'pkiss', 						{date => '20.03.2015', version => '2.2.12', name => 'pKiss', packageDir => 'pKiss/'},
-	'palikiss',					{date => '20.03.2015', version => '1.0.7', name => 'pAliKiss', packageDir => 'pAliKiss/'},
-	'libfoldgrammars', 	{date => '01.10.2015', version => '1.2.0', name => 'libfoldgrammars', packageDir => 'libfoldGrammars/'},
-	'rapidshapes', 			{date => '20.03.2015', version => '2.0.9', name => 'RapidShapes', packageDir => 'RapidShapes/'},
-	'knotinframe', 			{date => '20.03.2015', version => '2.0.8', name => 'knotinframe', packageDir => 'Knotinframe/'},
-	'rapidshapestest', 	{date => '01.03.2013', version => '2.1.0', name => 'RapidShapes-Test'},
-	'getoutsidetruth', 	{date => '19.04.2013', version => '1.0.0', name => 'getOutsideTruth'},
-	'acms',							{date => '21.03.2015', version => '1.2.1', name => 'acms', packageDir => 'aCMs/'},
-	'acmbuild', 					{date => '21.11.2014', version => '1.1.1', name => 'acmbuild'},
-	'acmsearch', 				{date => '21.11.2014', version => '1.1.1', name => 'acmsearch'},
+	'rnashapes', 				 {date => '18.11.2020', version => '3.3.1', name => 'RNAshapes', packageDir => 'RNAshapes/'},
+	'rnaalishapes', 		 {date => '18.11.2020', version => '2.4.8', name => 'RNAalishapes', packageDir => 'RNAalishapes/'},
+	'pkiss', 						 {date => '18.11.2020', version => '2.2.13', name => 'pKiss', packageDir => 'pKiss/'},
+	'palikiss',					 {date => '18.11.2020', version => '1.0.8', name => 'pAliKiss', packageDir => 'pAliKiss/'},
+	'libfoldgrammars', 	 {date => '01.10.2015', version => '1.2.0', name => 'libfoldgrammars', packageDir => 'libfoldGrammars/'},
+	'rapidshapes', 			 {date => '18.11.2020', version => '2.0.10', name => 'RapidShapes', packageDir => 'RapidShapes/'},
+	'knotinframe', 			 {date => '20.03.2015', version => '2.0.8', name => 'knotinframe', packageDir => 'Knotinframe/'},
+	'rapidshapestest', 	 {date => '01.03.2013', version => '2.1.0', name => 'RapidShapes-Test'},
+	'getoutsidetruth', 	 {date => '19.04.2013', version => '1.0.0', name => 'getOutsideTruth'},
+	'acms',							 {date => '21.03.2015', version => '1.2.1', name => 'acms', packageDir => 'aCMs/'},
+	'acmbuild', 				 {date => '21.11.2014', version => '1.1.1', name => 'acmbuild'},
+	'acmsearch', 				 {date => '21.11.2014', version => '1.1.1', name => 'acmsearch'},
 	'locomotif_wrapper', {date => '16.01.2015', version => '1.0.0', name => 'Locomotif_wrapper'},
 );
 
@@ -102,7 +102,7 @@ our $MODIFIER_UNKNOWN = 'unknown';
 our %RAPIDSHAPES_BIBISERV = (
 	'clusterwide_tempdir', '/vol/tmp/',
 	'tdmwrapper_binary', '/vol/fold-grammars/src/Misc/Applications/RapidShapes/tdmwrapper',
-	'binPath_grammargenerator', '/vol/fold-grammars/bin/', 
+	'binPath_grammargenerator', '/vol/fold-grammars/bin/',
 	'qsub', 'qsub -l virtual_free=6GB -l h_vmem=6GB -cwd -tc 10 ', #-tc controls the number of maximal parallel jobs for an array-job
 	'gridSH', '/usr/bin/sh',
 	'sleepTimeInit', '1', #initial number of seconds between two qstat requests
@@ -114,9 +114,9 @@ our %RAPIDSHAPES_BIBISERV = (
 my %checkedBinaries = (); #run time hash to avoid multiple check for binaries
 sub getBinary {
 	my ($requestedBinary) = @_;
-	
+
 	return $checkedBinaries{$requestedBinary} if (exists $checkedBinaries{$requestedBinary});
-	
+
 	my $binary = $requestedBinary;
 	$binary = $BINARIES{$requestedBinary} if (exists $BINARIES{$requestedBinary});
 	my $whichResult = qx(which $binary 2>&1);

